### PR TITLE
refactor: remove global semaphore

### PR DESCRIPTION
**Pull request by ash2k**
`rest.Config` has `RateLimiter` that should be used for rate limiting. It's not very good to have global state in a library too.

### CODE REVIEW

File: pkg/cache/cluster.go [hunk 1]
```diff
@@ -32,34 +31,6 @@ const (
 	ClusterRetryTimeout        = 10 * time.Second
 )
 
-var (
-	// listSemaphore is used to limit the number of concurrent k8s list queries.
-	// Global limit is required to avoid memory spikes during cache initialization.
-	// The default limit of 50 is chosen based on experiments.
-	listSemaphore = semaphore.NewWeighted(50)
```
Comment (alexmt) [inline pkg/cache/cluster.go:39]: Global limit on concurrent lists queries allows avoiding memory spike during cache initialization. Here is corresponding Argo CD ticket: https://github.com/argoproj/argo-cd/issues/3241 

@ash2k , is it possible to use `RateLimiter` to achieve a similar result?

Comment (ash2k) [reply pkg/cache/cluster.go:39]: @alexmt Hm, no, rate limiter does not allow to achieve what is needed. I've reworked what I'm suggesting to do, please see the commit comment.

Comment (alexmt) [reply pkg/cache/cluster.go:39]: sorry for delay. looking at PR today/tomorrow

File: pkg/cache/settings.go [hunk 2]
```diff
@@ -75,3 +75,25 @@ func SetConfig(config *rest.Config) UpdateSettingsFunc {
...
+
+// SetListPageBufferSize sets the number of pages to prefetch for list pager.
+func SetListPageBufferSize(listPageBufferSize int32) UpdateSettingsFunc {
+	return func(cache *clusterCache) {
+		cache.listPageBufferSize = listPageBufferSize
+	}
+}
+
+// SetListSemaphore sets the semaphore for list operations.
+// Taking an object rather than a number allows to share a semaphore among multiple caches if necessary.
```
Comment (alexmt) [inline pkg/cache/settings.go:94]: 👍

**Comment by ash2k** [comment]
@alexmt thanks for merging this, I appreciate it! I'll send some more PRs your way soon :)

**Pull request merged**